servlet/src/main/scala/org/http4s/servlet/AsyncHttp4sServlet.scala

```diff
@@ -66,9 +63,9 @@ class AsyncHttp4sServlet[F[_]](
       F.asyncF[Response[F]](cb => gate.complete(ctx.addListener(new AsyncTimeoutHandler(cb))))
     val response =
       gate.get *>
-        optionTSync
+        Sync[F]
           .suspend(serviceFn(request))
-          .getOrElse(Response.notFound)
+          .recover({ case _: scala.MatchError => Response.notFound[F] })
```

Comment:
It would be nice to be able to handle `NotFound` in a single place, that's why I used a `PartialFunction` and then handle the `NotFound` in the `recover`. Otherwise, when you have multiple services, you have to add `case _ => handleNotFound()` at the end of each of them. 

Furthermore, if you don't mount a service on the prefix `/`, then a request on anything caught by `/` will return an HTML that doesn't seem to be programmable.

Reply:
Sorry, I'm not quite sure what you're getting at here. I'll try and answer some of the points you are trying to make...

First, it looks like you've added this `recover` as a way to solve some issue in `JettyBuilder`, and it handles the situation that should not happen during any *normal* use of the `AsyncHttp4sServer`. This smells.

Second, in the Cats world we make a lot of assumptions, one of which is that all the functions are total except for `PartialFunction`s. We can only get a `MatchError` by breaking this rule via a non-exhaustive match - which, by the way, would lead to a compiler warning. Generally, any `MatchError` thrown signals broken code, and we should never try specifically handling it for an arbitrary function provided by the user code, never mind swallowing it without reporting. Basically, if you want to support a non-exhaustive `service` - use `HttpRoutes` instead. This is exactly what it is there for.

Third, regarding `case _ => handleNotFound()` - we can do `.orNotFound` when mounting `HttpRoutes` in `JettyBuilder`. I'm not sure why that should be a problem.

Finally, regarding `/`. As I understand the problem (if you can call it that) is that Jetty catches any requests that were not handled, and shows a standard 404 error page. I'm not sure it's our place to solve that by default, but even if it were - this could be solved by mounting a servlet to `/`. I'm still not sure what this has to do with the servlet implementation, though.

Reply:
Ok thank you for your comments. I think I understand your points. I will try to update my PR accordingly.

Reply:
@nigredo-tori I added `def orNotFound(notFoundHandler: A => Response[F]): Kleisli[F, A, Response[F]]` to `KleisliSyntax`. Also I updated `NotFoundHttpAppSpec`. It describes what I am trying to achieve here: having a simple, single way of handling custom `NotFound` responses accross the server. Do you think this would be reasonable ?

(also I agree with your latest comment about simply adding 
```
def mountHttpApp(service: HttpApp[F], prefix: String): Self =
    mountService(service.mapK(OptionT.liftK[F])), prefix)
```
to the builders)

Reply:
I've tried and summarized the important points of this below, and I suggest we continue this discussion there, since the scope of it went far beyond the single line of code I want gone. :smile: